### DISCUSSION TITLE
Doctrine Query Cache and readonly file systems

### DISCUSSION THREAD

**Discussion by umulmrum** [symfony/symfony#44656] [Q&A]
The DoctrineBundle config suggests using `cache.system` as query cache driver (see https://symfony.com/doc/current/reference/configuration/doctrine.html).

This means that system cache entries on the file system are generated during run time (as opposed to build time), which of course is not possible in readonly file systems. Is there a recommendation what to do in this case? Is it a good idea to use `cache.app` with Redis and array cache chain? Thanks!

**Comment by gizmola** [comment]
This advice is for the query and metadatacache only, but I would consider APCu.  To be clear, at one point APC was a PHP opcode cache AND a user cache api, but the PHP opcode cache features were removed by the developers who created APCu.  It makes use of a user defined amount of shared memory so in that way, if you have a cluster, you are distributing the memory and storage requirements across your cluster, which reduces the number of single points of failure.  Reading and writing to shared memory is going to be very fast.  For the result cache, using memcache or Redis makes sense, assuming you have code to make use of the cache in an intelligent fashion.

**Reply by umulmrum** [reply]
@gizmola Thanks for your answer. I hesitate to move the cache away from files as I suppose that it won't be grabbed by OpCache and therefore hit performance - right? Are there any docs/blog posts etc. where caching options are discussed? It is hard to find info on the details.

**Reply by gizmola** [reply]
Caching to a read-only fs is not an option for you, correct?  You will continue to use OpCache and it will accelerate what it can.  You would only use the apcu cache driver for the meta and query caches which you can't cache to files.  Each of the 3 possible Doctrine caches can be configured independently.  Between Redis and apcu, apcu is implemented with shared memory on your machine, so this is highly performant, and introduces no network overhead.  The cache requirements for the doctrine meta and query cache are typically very small.  For the result cache, I would use memcached or redis, but again, you have to write code to utilize a result cache, as only you know for your application what is appropriate data to cache, and for how long.

**Reply by umulmrum** [reply]
OK, thank you. I already use Redis as result cache, and metadata cache can stay on the file system as it is pregenerated in the CI/CD pipeline. I will try APCu for the query cache and check if it's viable :-)

**Reply by gizmola** [reply]
If you already do plan to use Redis, and you will be pre-generating the meta cache, then I would probably just use Redis for the query and result cache.
